instance/models/server.py

```diff
@@ -0,0 +1,274 @@
+            return None
+
+        return public_addr['addr']
+
+    def update_status(self, provisioned=False, rebooting=False):
+        """
+        Refresh the status by querying the openstack server via nova
+        """
+        # TODO: Check when server is stopped or terminated
+        os_server = self.os_server
+        self.log('debug', 'Updating status for {} from nova (currently {}):\n{}'.format(
+            self, self.status, to_json(os_server)))
+
+        if self.status == 'started':
+            self.log('debug', 'Server {}: loaded="{}" status="{}"'.format(
+                self, os_server._loaded, os_server.status))
+            if os_server._loaded and os_server.status == 'ACTIVE':
+                self._set_status('active')
+
+        elif self.status == 'active' and is_port_open(self.public_ip, 22):
+            self._set_status('booted')
+
+        elif self.status == 'booted' and provisioned:
+            self._set_status('provisioned')
+
+        elif self.status in ('provisioned', 'ready') and rebooting:
+            self._set_status('rebooting')
+
+        elif self.status == 'rebooting' and not rebooting and is_port_open(self.public_ip, 22):
+            self._set_status('ready')
+
+        return self.status
+
+    def start(self):
+        """
+        Get a server instance started and an openstack_id assigned
+
+        TODO: Add handling of quota limitations & waiting list
+        TODO: Create the key dynamically
+        """
+        self.log('info', 'Starting server {} (status={})...'.format(self, self.status))
+        if self.status == 'new':
+            os_server = openstack.create_server(
+                self.nova,
+                self.instance.sub_domain,
+                settings.OPENSTACK_SANDBOX_FLAVOR,
+                settings.OPENSTACK_SANDBOX_BASE_IMAGE,
+                key_name=settings.OPENSTACK_SANDBOX_SSH_KEYNAME,
+            )
+            self.openstack_id = os_server.id
+            self.log('info', 'Server {} got assigned OpenStack id {}'.format(self, self.openstack_id))
+            self._set_status('started')
+        else:
+            raise NotImplementedError
+
+    def reboot(self, reboot_type='SOFT'):
+        """
+        Reboot the server
+
+        This requires to switch the status to 'rebooting', which is first attempted via the
+        `update_status` method. If the current state doesn't allow to switch to this status,
+        a ServerNotReady exception is thrown.
+        """
+        if self.update_status(rebooting=True) != 'rebooting':
+            raise ServerNotReady("Can't change status to 'rebooting' (current: '{}')".format(self.status))
+        self.os_server.reboot(reboot_type=reboot_type)
+
+        # TODO: Find a better way to wait for the server shutdown and reboot
+        # Currently, without sleeping here, the status would immediately switch back to ready,
+        # as SSH is still available until the reboot terminates the SSHD process
```

Comment:
@antoviaque What about sending a command like `bash -c "sleep 5; shutdown -r now" & service ssh stop` that will synchronously stop the SSH server and then asynchronously (but deterministically) reboot the server? (The single ampersand between those commands is intentional.) As soon as that command has completed, you can wait for port 22 to open again and then you know that the server has rebooted.